internal/customgitignore/walk_up_to_root.go

```diff
@@ -0,0 +1,247 @@
+package customgitignore
+
+import (
+	"errors"
+	"os"
+	"path/filepath"
+	"strings"
+
+	"github.com/go-git/go-billy/v5"
+	"github.com/go-git/go-billy/v5/osfs"
+	"github.com/go-git/go-git/v5"
+	"github.com/go-git/go-git/v5/plumbing/format/gitignore"
+)
+
+// Takes a dir and processes .gitignore files from it.
+//
+// This uses go-git under the hood and returns a slice
+// of go-git's gitignore.Pattern structs.
+//
+// It also returns the path to the root of the git-repo,
+// or "" if this path isn't within a git repo, allowing
+// a caller to know how .gitingore files were parsed.
+//
+// The actual parsing is inteded to be similar to how tools
+// like rg work, but means that `path` may not necessarily be
+// the root of a git repo, and can produces these parsing
+// behaviours:
+//
+// `path` is a plain dir:
+//
+//   - .gitignore files are ignored
+//
+// `path` is a file:
+//
+//   - find the file's dir and then run the following rules...
+//
+// `path` is a dir at the root of a git-repo, with recursive flag:
+//
+//   - read all .gitignore files in repo
+//   - read .git/info/exclude for repo
+//
+// `path` is a dir at the root of a git-repo, without recursive flag:
+//
+//   - only read .gitignore files in repo-root
+//   - read .git/info/exclude for repo
+//
+// `path` is a dir within a git-repo, with recursive flag:
+//
+//   - read .gitignore in start dir
+//   - read all .gitignore files in child-dirs
+//   - read all .gitignore files in parent dirs up to and including repo-root
+//     (but only dirs that are an ancestor)
+//
+// `path` is a dir within, a git-repo, without recursive flag:
+//
+// 		- read .gitignore in start dir
+// 		- read all .gitignore files in parent dirs up to and including repo-root
+// 			(but only dirs that are an ancestor)
+// 		- read .git/info/exclude for repo
+//
+// NOTE: the dir you're passing in directly could be a dir that is ignored
+// (targeted by a parent's .gitignore or the per-repo exclude file); in this
+// case, the dir's .gitignore file is still processed, but not its sub-dirs.
+
```

Comment:
I think this also needs a `//` otherwise go will think this is two different comment blocks.
```suggestion
//
```

Reply:
good catch. godoc would have made a small, and rather confusing summary for that func.